src/tests/timer-add-api-test/timer-add-api-test.c

```diff
@@ -0,0 +1,189 @@
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+
+#include "common_types.h"
+#include "osapi.h"
+#include "utassert.h"
+#include "uttest.h"
+#include "utbsp.h"
+
+#define NUMBER_OF_TIMERS 4
+#define TASK_1_ID         1
+#define TASK_1_STACK_SIZE 4096
+#define TASK_1_PRIORITY   101
+
+OS_time_t        StartTime;
+OS_time_t        EndTime;
+uint32           TimerStart[NUMBER_OF_TIMERS] = {1000, 2000000, 3000000, 4000000 };
+uint32           TimerInterval[NUMBER_OF_TIMERS] = {500000, 400000, 800000, 600000 };
+
+uint32 TimerTestTaskStack[TASK_1_STACK_SIZE];
+int32 timer_counter[NUMBER_OF_TIMERS];
+uint32 timer_idlookup[OS_MAX_TIMERS];
+
+/*
+** Test timer function.
+** Note: For some Host OSs, this is the equivalent of an ISR, so the calls available are limited.
+** For example, Linux and vxWorks can call functions like printf, but RTEMS cannot.
+*/
+void test_func(uint32 timer_id , void *arg)
+{
+   OS_ConvertToArrayIndex(timer_id, &timer_id);
+   timer_counter[timer_idlookup[timer_id]]++;
+}
+
+/* *************************************** MAIN ************************************** */
+
+void TestTimerAddApi(void)
+{
+    /*
+     * Test Case For:
+     * int32 OS_TimerAdd(uint32 *timer_id, const char *timer_name, uint32 timebase_ref_id, OS_ArgCallback_t  callback_ptr, void *callback_arg)
+     */
+
+    int32    actual;
+    int32    expected;
+    uint32   timer_id;
+    char     arg = 'a';
+    uint32   time_base_id;
+    int      i = 0;
+    int32    TimerStatus[NUMBER_OF_TIMERS];
+    uint32   TableId;
+    uint32   TimerID[NUMBER_OF_TIMERS];
+    char     TimerName[NUMBER_OF_TIMERS][20] = {"TIMER1","TIMER2","TIMER3","TIMER4"};
+    uint32   microsecs;
+
+    OS_TimeBaseCreate( &time_base_id, "TimeBase", 0);
+    OS_TimeBaseSet(time_base_id, 10000, 10000); //ms
+
+   for ( i = 0; i < NUMBER_OF_TIMERS; i++ )
+   {
+      TimerStatus[i] = OS_TimerAdd(&TimerID[i], TimerName[i], time_base_id, &test_func, &arg);
```

Comment:
The "arg" passed here as the last parameter should be a pointer to the counter you want to increment.  The old TimerCreate API didn't have this, which is why it had to do a workaround based on the ID.

This is the other big difference between old TimerCreate and new TimerAdd API, in that it has this extra data object pointer.  This test should show it being used for its intended purpose.

Reply:
So the last parameter here should be `&timer_counter[i]`